src/server/pkg/log/log.go

```diff
@@ -26,16 +30,108 @@ func NewLogger(service string) Logger {
 	l.Formatter = new(prettyFormatter)
 	return &logger{
 		l.WithFields(logrus.Fields{"service": service}),
+		make(map[string]*prometheus.HistogramVec),
+		make(map[string]prometheus.Counter),
 	}
 }
 
 // Helper function used to log requests and responses from our GRPC method
 // implementations
 func (l *logger) Log(request interface{}, response interface{}, err error, duration time.Duration) {
+	state := "started"
 	if err != nil {
 		l.LogAtLevelFromDepth(request, response, err, duration, logrus.ErrorLevel, 4)
+		state = "errored"
 	} else {
 		l.LogAtLevelFromDepth(request, response, err, duration, logrus.InfoLevel, 4)
+		if duration.Seconds() > 0 {
+			state = "finished"
+		}
+	}
+	l.ReportMetric(state, duration)
+}
+
+func (l *logger) ReportMetric(state string, duration time.Duration) {
+	depth := 4
+	pc := make([]uintptr, depth)
+	runtime.Callers(depth, pc)
+	split := strings.Split(runtime.FuncForPC(pc[0]).Name(), ".")
+	method := split[len(split)-1]
+	entry := l.WithFields(logrus.Fields{"method": method})
+
+	var tokens []string
+	for _, token := range camelcase.Split(method) {
+		tokens = append(tokens, strings.ToLower(token))
+	}
+	rootStatName := strings.Join(tokens, "_")
+
+	// Recording the distribution of started times is meaningless
+	if state != "started" {
+		runTimeName := fmt.Sprintf("%v_time", rootStatName)
+		runTime, ok := l.histogram[runTimeName]
```

Comment:
This seems unsafe, what if 2 calls to `ReportMetrics` are running concurrently? Won't this be a concurrent map access?

Reply:
Yea :-/ ... I'll put in a mutex to fix that

Reply:
[Actually, seems like this is one of the 2 recommended use cases for the new conccurent map](https://github.com/golang/go/blob/master/src/sync/map.go#L20)

So I'll use that instead, unless we have a preference against that new primitive

Reply:
My only preference against it is that it's a bit clunky to use compared to real maps, but it's not that bad. Either approach is fine imo, although if you do want to use a Mutex I think it's probably worth it to use an RWMutex since these maps are going to mostly experience reads.

Reply:
Just went w a mutex, seemed easier

Reply:
So I was just reading through this and I think it all makes sense. The one thing I'm worried about is if you get a lot of log requests then you might get a growing backlog of goros waiting on this mutex. I think it's fine because I think what happens in this function should be very fast, so it shouldn't really be possible to get a backlog. How would you feel about adding a `gauge` for the number of active `ReportMetric` calls we have going? That way at least if this is a problem we'll be able to see it in our metrics.

Reply:
Yea I was wondering how we handled goro's that we spawn and forget. I like this.